### DISCUSSION TITLE
How would you handle errors thrown from the authedProcedure checks?

### DISCUSSION THREAD

**Discussion by alexnaish** [trpc/trpc#3087] [Q&A] [answered]
First of all, I just want to say that I'm thoroughly enjoying developing using tRPC and I'm blown away by the work and support involved with it!

In your authorisation docs, you have an example of a procedure that uses a middleware to check a context value. When this case  does throw an `UNAUTHORIZED` error - where / how would you suggest handling it? 

I'm leaning towards attaching an onError handler to my queryClientConfig - upon receiving an error, check the message / code and handle appropriately (most likely `window.location.href = 'login page here'` or similar). However although the default onError handle is typed to receive an `unknown` error, the handler in this situation receives a `TRPCClientError` which has `data` and `shape` properties which contain relevant properties for my use case.

I'm thinking to add something like the below to my `createTRPCNext<AppRouter>` config section:

```
queryClientConfig: {
  defaultOptions: {
    queries: {
      onError: (error: any) => {
        if (error?.data?.code === 'UNAUTHORIZED') {
          window.location.href = 'login page here';
        }
      },
    }
  }
}
```

What are your thoughts?

**Comment by KATT** [comment, marked as answer]
I've seen different solutions to this:

- You can use a custom [link](https://trpc.io/docs/v10/links) and display a toast or something 
- You can handle it inline on a per-query basis
- You can handle it globally as you're suggesting above
- You can have a global app provider that does it

Personally, my React-tree has a root `<AppProvider>` where I always query `viewer.me` ahead of going further down the tree (which will 401 if you're not authed), and in there I forward the user to the login page. This will have a query waterfall of subsequent queries, but I don't really mind that.
